RecoBTag/FeatureTools/plugins/ParticleNetFeatureEvaluator.cc

```diff
@@ -0,0 +1,807 @@
+  desc.add<edm::InputTag>("pf_candidates", edm::InputTag("packedPFCandidates"));
+  desc.add<edm::InputTag>("losttracks", edm::InputTag("lostTracks"));
+  desc.add<edm::InputTag>("jets", edm::InputTag("slimmedJetsAK8"));
+  desc.add<edm::InputTag>("muons", edm::InputTag("slimmedMuons"));
+  desc.add<edm::InputTag>("taus", edm::InputTag("slimmedTaus"));
+  desc.add<edm::InputTag>("electrons", edm::InputTag("slimmedElectrons"));
+  desc.add<edm::InputTag>("photons", edm::InputTag("slimmedPhotons"));
+  descriptions.add("ParticleNetFeatureEvaluator", desc);
+}
+
+void ParticleNetFeatureEvaluator::produce(edm::Event &iEvent, const edm::EventSetup &iSetup) {
+  // output collection
+  auto output_tag_infos = std::make_unique<std::vector<reco::DeepBoostedJetTagInfo>>();
+  // Input jets
+  auto jets = iEvent.getHandle(jet_token_);
+  // Input muons
+  auto muons = iEvent.getHandle(muon_token_);
+  // Input taus
+  auto taus = iEvent.getHandle(tau_token_);
+  // Input electrons
+  auto electrons = iEvent.getHandle(electron_token_);
+  // Input photons
+  auto photons = iEvent.getHandle(photon_token_);
+  // Input lost tracks
+  iEvent.getByToken(losttrack_token_, losttracks_);
+  // Primary vertexes
+  iEvent.getByToken(vtx_token_, vtxs_);
+  if (vtxs_->empty()) {
+    // produce empty TagInfos in case no primary vertex
+    iEvent.put(std::move(output_tag_infos));
+    return;  // exit event
+  }
+  // Leading vertex
+  pv_ = &vtxs_->at(0);
+  // Secondary vertexs
+  iEvent.getByToken(sv_token_, svs_);
+  // PF candidates
+  iEvent.getByToken(pfcand_token_, pfcands_);
+  // Track builder
+  track_builder_ = iSetup.getHandle(track_builder_token_);
+
+  // tau signal candidates
+  std::vector<math::XYZTLorentzVector> tau_pfcandidates;
+  for (size_t itau = 0; itau < taus->size(); itau++) {
+    if (taus->at(itau).pt() < min_pt_for_taus_)
+      continue;
+    if (fabs(taus->at(itau).eta()) > max_eta_for_taus_)
+      continue;
+    for (unsigned ipart = 0; ipart < taus->at(itau).signalCands().size(); ipart++) {
+      const pat::PackedCandidate *pfcand =
+          dynamic_cast<const pat::PackedCandidate *>(taus->at(itau).signalCands()[ipart].get());
+      tau_pfcandidates.push_back(pfcand->p4());
+    }
+  }
+
+  // Loop over jet
+  for (std::size_t jet_n = 0; jet_n < jets->size(); jet_n++) {
+    const auto &jet = (*jets)[jet_n];
+    edm::RefToBase<reco::Jet> jet_ref(jets, jet_n);
+
+    // create jet features
+    DeepBoostedJetFeatures features;
+    for (const auto &name : particle_features_)
+      features.add(name);
+    for (const auto &name : sv_features_)
+      features.add(name);
+
+    // fill values only if above pt threshold and has daughters, otherwise left
+    bool fill_vars = true;
+    if ((jet.pt() < min_jet_pt_ and
+         dynamic_cast<const pat::Jet *>(&jet)->correctedJet("Uncorrected").pt() < min_jet_pt_) or
+        std::abs(jet.eta()) > max_jet_eta_ or std::abs(jet.eta()) < min_jet_eta_)
```

Comment:
This is an edge case, but I think right now the features are filled twice (both central and forward taggers) for a jet with an eta of exactly 2.5. Given that the forward tagger is only used (e.g. in Nano) for `<2.5`, perhaps this could be reflected here? (ie change to `>= max_jet_eta_`?)

Reply:
Sure. For the record, the central jet training requires eta < 2.5 and the forward jet training eta > 2.5, so technically a jet with eta==2.5 (I guess very very rare given the float precision) is just as well defined to be used by either training. I would propose that it get the central network evaluation, so <= min_jet_eta and > max_jet_eta in this line.

Reply:
Shouldn't it be `>= max_jet_eta` and `< min_jet_eta`, since you then set `fill_vars = false`?

With `<= min_jet_eta`, jets with eta==0.0 won't get the variables filled.

Though as you say, with float precision is should be very rare anyway...